Replace age field with DOB in users model

rash-27 opened the pull request:
## Proposed Changes

- Replace the age field with the date of birth in Users model 

### Associated Issue

-  #7281 in the frontend requires this change 

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

rithviknishad reviewed (changes requested):
Missing custom migration to migrate existing users' age to DOB

rash-27 commented:
Ok I'll update it

sainak commented on care/users/api/serializers/user.py:
```diff
@@ -323,6 +332,12 @@ class Meta:
     extra_kwargs = {"url": {"lookup_field": "username"}}
 
+    def validate_date_of_birth(self, value):
+        if value and datetime.today().year - value.year <= 0:
+            raise serializers.ValidationError("Enter a valid date of birth.")
```
```suggestion
        if value and now().year - value.year < 16:
            raise serializers.ValidationError("Age must be greater than 15 years")
```

sainak reviewed (changes requested):
update dummy data

sainak commented on care/users/models.py:
```diff
@@ -235,7 +234,7 @@ class User(AbstractUser):
 
     gender = models.IntegerField(choices=GENDER_CHOICES, blank=False)
-    age = models.IntegerField(validators=[MinValueValidator(1), MaxValueValidator(100)])
+    date_of_birth = models.DateField(null=True, blank=True)
```
make this non nullable after migration

sainak commented on care/users/api/serializers/user.py:
```diff
@@ -112,6 +111,14 @@ class Meta:
             "user_permissions",
             "created_by",
         )
+    date_of_birth = serializers.DateField(required=False)
```
```suggestion
    date_of_birth = serializers.DateField(required=True)
```

sainak commented on care/users/api/serializers/user.py:
```diff
@@ -275,6 +282,8 @@ class UserSerializer(SignUpSerializer):
 
     home_facility = ExternalIdSerializerField(queryset=Facility.objects.all())
 
+    date_of_birth = serializers.DateField(required=False)
```
```suggestion
    date_of_birth = serializers.DateField(required=True)
```

rithviknishad commented on care/users/migrations/0015_age_to_dateofbirth.py:
```diff
@@ -0,0 +1,35 @@
+            field=models.DateField(blank=True, null=True),
+        ),
+
+        migrations.RunPython(age_to_date_of_birth, migrations.RunPython.noop),
```
Needs reverse_code to reverse DOB to age

vigneshhari commented:
Does it make sense to use Jan 1 in the migrations, a random DOB seems wrong to me.

rash-27 commented:
@vigneshhari here in the database we store only the age previously in Users model, 
1. Choosing a random value of date of birth is  to save the age of the user as we eliminate the age field after migration 
2. choosing only jan 1 as random dob is because , we can accurately calculate the age if needed , if we choose another date , our age calculated will get inconsistent if the actual date of birth is before the chosen date of birth
Please correct me , if I am wrong

vigneshhari commented:
okays, we can merge after the lint and tests pass

rash-27 commented:
@vigneshhari I have modified the tests and they are passing and lint errors are also fixed ,once please review it 
![image](https://github.com/coronasafe/care/assets/146672184/7416c2f6-2d61-4431-abde-52b83ca6aa9e)
